File: top150/_02_TwoPointers/q11.java
package top150.TwoPointers;

public class q11 {
    public int maxArea(int[] height) {
        int head = 0;
        int tail = height.length - 1;

        int max = 0;
        while(head < tail) {
            max = Math.max(max, Math.min(height[head], height[tail]) * (tail-head));
            if(height[head] > height[tail]) tail--;
            else head++;
        }

        return max;

    }
}

File: top150/_02_TwoPointers/q11.py
import unittest


class Solution:
    # T: O(N), S: O(1)
    def maxArea(self, height):
        result = 0
        n = len(height)
        l = 0
        r = n-1
        while l < r:
            if height[l] < height[r]:
                result = max(result, height[l]*(r-l))
                l += 1
            else:
                result = max(result, height[r]*(r-l))
                r -= 1
        return result


class Test(unittest.TestCase):
    def test_one(self):
        nums = [1, 8, 6, 2, 5, 4, 8, 3, 7]
        sol = Solution()
        result = sol.maxArea(nums)
        self.assertEqual(result, 49)

    def test_two(self):
        nums = [1, 1]
        sol = Solution()
        result = sol.maxArea(nums)
        self.assertEqual(result, 1)


# entry point
if __name__ == '__main__':
    unittest.main()
